Cleanup & Fixes: TIO, Keys, Beats, TrackDAO

uklotzde opened the pull request:
Lots of minor cleanups, fixes and improvements around the mentioned classes. Some of the modifications depend on others so I combined them in a single PR.

uklotzde committed:
TIO: Improve const-correctness

uklotzde committed:
TIO: Delete obsolete code (already commented out)

uklotzde committed:
Reduce risks for memory leaks in Keys interface

uklotzde committed:
Reduce risks for memory leaks in Beats interface

uklotzde committed:
Simplify and optimize the over-engineered Keys class

This class should be a simple value object class:
- The class was never intended as a base class
- The thread synchronization was incomplete and unnecessary

uklotzde committed:
TrackDAO: Replace plain pointers with std::unique_ptr

uklotzde committed:
TrackDAO: Merge code for binding DB columns

uklotzde committed:
Consolidate code for extracting embedded cover art

uklotzde committed:
TrackDAO: Updating tracks in library may fail

Fixes a debug assertion when switching library directories

daschuer commented on src/track/beatgrid.cpp:
```diff
@@ -67,27 +67,26 @@ void BeatGrid::setGrid(double dBpm, double dFirstBeatSample) {
-    QByteArray* pByteArray = new QByteArray(output.data(), output.length());
-    // Caller is responsible for delete
-    return pByteArray;
+    return std::make_unique<QByteArray>(output.data(), output.length());
```
I think we can improve this even more. 
This function is only used here 
https://github.com/mixxxdj/mixxx/blob/87ef849b94e6080b307fff33793567682c0687a2/src/library/dao/trackdao.cpp#L1605
To finally create a temorary QVariant. 

How about
```
QVariant BeatGrid::toVariant() const {
     QMutexLocker locker(&m_mutex);
     std::string output;
     m_grid.SerializeToString(&output);
     return QVariant(output.c_str);
}
```

daschuer commented on src/track/beatgrid.cpp:
```diff
@@ -245,18 +244,18 @@ bool BeatGrid::findPrevNextBeats(double dSamples,
+        return std::unique_ptr<BeatIterator>();
     }
-    return new BeatGridIterator(m_dBeatLength, curBeat, stopSample);
+    return std::make_unique<BeatGridIterator>(m_dBeatLength, curBeat, stopSample);
```
It feels weird to delete an Iterator, but I understand that how it currently works. 

It is only used here
https://github.com/mixxxdj/mixxx/blob/49aaf97de17612b57d2a7618fcae0e687dbea8e0/src/waveform/renderers/waveformrenderbeat.cpp#L52
Cant we return the BeatGridIterator by value and rely on the Compilers return value optimization that avoids a deep copy?

uklotzde replied:
BeatIterator is an abstract class and could not returned by value. The 
interface is defined by the base class.

On 02/11/2016 09:06 AM, Daniel Schürmann wrote:
In src/track/beatgrid.cpp 
<https://github.com/mixxxdj/mixxx/pull/892#discussion_r52571750>:
> >      }
> > -    return new BeatGridIterator(m_dBeatLength, curBeat, stopSample);
> > +    return std::make_unique<BeatGridIterator>(m_dBeatLength, curBeat, stopSample);
>
> It feels weird to delete an Iterator, but I understand that how it 
> currently works.
>
> It is only used here
> https://github.com/mixxxdj/mixxx/blob/49aaf97de17612b57d2a7618fcae0e687dbea8e0/src/waveform/renderers/waveformrenderbeat.cpp#L52
> Cant we return the BeatGridIterator by value and rely on the Compilers 
> return value optimization that avoids a deep copy?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/892/files#r52571750>.
>

uklotzde committed:
Simplify serialization of Beats and Keys

QByteArray uses implicit sharing and does not need to be allocated
on the heap.

daschuer commented:
Thank you for another nice clean up. LGTM
